sys/arduino/base.cpp

```diff
@@ -106,3 +107,50 @@ int analogRead(int arduino_pin)
     return adc_value;
 }
 #endif
+
+#if MODULE_PERIPH_PWM
+#define PWM_MODE            PWM_LEFT
+#define PWM_FREQU           (1000U)
+#define PWM_STEPS           (256U)
+
+static int _get_pwm_pin_idx(int pin)
+{
+    for (uint8_t i = 0;i < ARRAY_SIZE(arduino_pwm_list); ++i) {
+        if (arduino_pwm_list[i].pin == pin) {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+void analogWrite(int pin, int value)
+{
+    /*
+    * Bitfield for the state of the PWM-channels.
+    * 0: Not initialized
+    * 1: Successfully initialized
+    */
+    static uint16_t pwm_dev_state;
```

Comment:
I fear its not quite that simple: If one would call a digital_write() on that same pin in the mean time, the bit value in this variable is not valid anymore. So it seems to me, that this value must be shared for all functions in this module and it must be also updated by other functions using these pins...

Reply:
Not sure, here it's only used to check if the PWM device is already initialized. It has nothing to do with the pin itself.

Reply:
but `pwm_init` also takes care of configuring the output state/mode for the effected pin. So maybe this might be a more generic problem. What happens for code like the following in Arduino?
```c
pinMode(X, OUTPUT);
digitalWrite(X, HIGH);
analogWtrite(X, 100);
pinMode(X, OUTPUT);
digitalWrite(X, HIGH);
analogWrite(X, 50);
```
Here one would expect, that the wave form at pin X is active after the last `analogWrite` call. Now the bitfield above will actually prevent that from happen, as it thinks the pin is already initialized as PWM output, which is not the case anymore...

Reply:
You are right. Looking at the [samd21 implementation of analogWrite](https://github.com/arduino/ArduinoCore-samd/blob/aef4af4c1e20ffeee2a8e62c6998d1fcb2c7fa4b/cores/arduino/wiring_analog.c#L194) in Arduino, it seems the is reinitialized all time but not the TCC. This is a design incompatibility between Arduino and RIOT here.

Reply:
Considering that Arduino is not really geared towards maximum effenciency, I think we should be good with re-initializing the PWM on every call, right? IMHO API complience is more important than efficiency considerations.

Only question is how to/if we need to deal with signal ripples at the pin that could potentially be caused by this re-initialization. But the exact same behavior is triggered on the very first call in any case, so we might just be able to ignore that issue.

Reply:
Looking at the RIOT `pwm_init` function, I don't understand why this function initializes all pins configured on a given PWM device. If another pin is used for another function, then the order of initialization might break it.
Introducing the `pwm_gpio_init` could solve this issue, even on the RIOT side. And we could remove the calls to `pwm_gpio_init` in `pwm_init`.

Reply:
This is once more a problem of how the periph APIs were designed and of how people started to use them: in the initial design we always assumed a static 1-to-1 mapping of pins to periph units. So initializing all **configured** PWM pins does not hurt, as they should not be used somewhere else. Works perfectly fine for 99% of production boards, but of course hits some limitations for non-specified eval platforms, which include the usage of the Arduino API.

So simply changing the `periph/pwm` API does not solve this issue...

Reply:
Additional questions regarding `pwm_gpio_init`.

- Is this function really necessary or wouldn't it be possible to initialize the according pin when the channel is set the first time?
- If the function is necessary, shouldn't it be part of the PWM API. It's a bit weird that then name of a function in the GPIO API starts with `pwm`.
- Have you already checked which of the architectures that support the Arduino API require would require that function. For example, ESP32 also initializes the pins of all channesl during `pwm_init`.

Another question is, what if the application tries to use the same PWM device for PWM channels while `analogWrite` is used? The relation between PWM and `analogWrite` is not obvious.

Reply:
> Is this function really necessary or wouldn't it be possible to initialize the according pin when the channel is set the first time?

`pwm_init` initializes all channels on all CPU specific implementations (at least on the ones I checked). I don't think this is necessary but this is how it's done in RIOT.
I would tend to add a `pwm_dev_init` that would only reset and initialize the PWM device but not any channels. The existing `pwm_init` could stay in the API and to the same job: call `pwm_dev_init` and `pwm_channel_init` on all configured channels.

> If the function is necessary, shouldn't it be part of the PWM API. It's a bit weird that then name of a function in the GPIO API starts with pwm.

It's added in the pwm API but maybe the name is not good. It could be called `pwm_channel_init` instead ?

> Have you already checked which of the architectures that support the Arduino API require would require that function.

I checked STM32, AVR and samd21. All are doing the same. So this function (`pwm_gpio/channel_init`) could be defined there as well.

> what if the application tries to use the same PWM device for PWM channels while analogWrite is used?

This might be possible I guess, if it doesn't change the settings of the device. But the question is more if a pin configured on a PWM channel can be used for something else (digitalWrite), if another channel of the same PWM is used in analogWrite. For the moment, with the current implementation, it depends on the order of calling the functions:
```cpp
pinMode(x OUTPUT);
digitalWrite(x, HIGH);
analogWrite(y, 128);
digitalWrite(x, LOW);
```
This won't work if X and Y are 2 pins configured for 2 channels on the same PWM device because they are reinitialized for PWM in the first call to analogWrite. The second call to digitalWrite might not work (untested but with the current implementation that should be the case).

Reply:
>> If the function is necessary, shouldn't it be part of the PWM API. It's a bit weird that then name of a function in the GPIO API starts with pwm.

> It's added in the pwm API but maybe the name is not good.

Yes, of course :confused: Sorry, my faul. For any reason I saw it in `gpio.h`.

> It could be called `pwm_channel_init` instead ?

Not sure, while `pwm_gpio_init` says what it does, `pwm_channel_init` would be more consistent with terminology in `pwm.h`. Maybe, `pwm_channel_init` would be better.

Reply:
```
pinMode(x OUTPUT);
digitalWrite(x, HIGH);
analogWrite(y, 128);
digitalWrite(x, LOW);
```
> This won't work if X and Y are 2 pins configured for 2 channels on the same PWM device because they are reinitialized for PWM in the first call to `analogWrite`. The second call to `digitalWrite` might not work (untested but with the current implementation that should be the case).

It could work with your proposed change to split the functionality of `pwm_init` Into `pwm_dev_init` and `pwm_channel_init`. What if PWM channels were never initialized in `pwm_init`, but always in `pwm_set` when a certain channel is used for the first time?

Reply:
> What if PWM channels were never initialized in pwm_init, but always in pwm_set when a certain channel is used for the first time?

Doing this will make `pwm_set` slightly less efficient but it will also be a little more robust (if the same pin is initialized differently between 2 calls to pwm_set.

May I suggest to drop the refactoring proposal from this PR and work on this separately ?

Reply:
> Doing this will make `pwm_set` slightly less efficient

Yes, but how often is it called in real applications? I woud guess not so often so that this overhead seems to be acceptable.

> May I suggest to drop the refactoring proposal from this PR and work on this separately ?

Sure.

Reply:
> but how often is it called in real applications? I woud guess not so often so that this overhead seems to be acceptable.

I tend to agree here. And it's just 2 extra function calls. Doing this change, e.g not initialize all channels in pwm_init and only initialize one channel on demand in pwm_set, would also save 52 bytes on samd21...